### ISSUE TITLE
Make Scalar-valued attributes work in constraints

### ISSUE THREAD

**Issue by alberts** [mesosphere/marathon#1711]
Mesos 0.22.1 attributes that have integer values don't work if you try to use them as constraints in Marathon 0.8.2. Jobs never get scheduled.

Not quite sure where things go wrong.

**Comment by pierlo-upitup** [comment]
Hey @alberts,

are you sure the attribute is defined on the slave?

> If the specified attribute is not defined on the slave, then most operators will refuse to run tasks on this slave. In fact, only the UNLIKE operator will (and always will) accept this offer for now, while other operators will always refuse it.

See :
https://github.com/mesosphere/marathon/blob/master/docs/docs/constraints.md#attribute-field

Could you post the relevant parts of your JSON definition for the app you are trying to deploy?

Thanks.

**Comment by bobrik** [comment]
This is how mesos slave reports in `/state` of mesos master api:

```
    {
      "active": true,
      "attributes": {
        "rack": 109,
        "type": "compute"
      },
      "hostname": "myhost",
```

This is how constraint is specified:

```json
[["rack","CLUSTER","109"]]
```

Marathon fails to schedule tasks in this case.

Mesos 0.25.0, Marathon 0.11.1.

**Comment by kolloch** [comment]
This is a known limitation that is documented here (https://mesosphere.github.io/marathon/docs/constraints.html):

"""
Note that currently Marathon only supports text values for attributes and will not be able to process any numberic values. For example the attribute pair foo:bar will be recognised, but marathon will not be able to match cpu:4.
"""

That doesn't make it pretty, though.

I would like to keep this issue as a reminder to extend the functionality.

**Comment by bobrik** [comment]
@kolloch do you have any plans for using numeric constraints differently? I thought that attributes are opaque strings, looks like I could be wrong.

**Comment by kolloch** [comment]
Attributes have the same types as resources but they are not consumed. Thus if the value has the "shape" of an integer, Mesos will use a different type for them. See [attributes/resources in Mesos](http://mesos.apache.org/documentation/attributes-resources/)

**Comment by bobrik** [comment]
Any chance to at least pretend that numeric attributes are strings before implementing full support for numbers and ranges?

This way you can have your `rack:123` attribute now and group by it. Once numbers land, you'll be able to use ranges, if you wish. All of that without changing attributes to hacks like `rack:r123` and wiping mesos slave state twice.

**Comment by kolloch** [comment]
If I am not mistaken, treating all values as strings should be ridiculously easy. I would have to think about whether that's always what you want (for ranges it's really strange) -- but we simply could introduce separate operators for scalar types where needed.

Can you explain why this is important to you? Right now, the biggest problem in my view is that things don't work as expected but you can work around them once you got why.

**Comment by bobrik** [comment]
Restarting slaves and losing their tasks is the problem I have. I don't wan't to do that. I already have `rack:123` attributes in place and I want to use them.

**Comment by abraithwaite** [comment]
@alberts 🦄🚀👾!

**Comment by brndnmtthws** [comment]
This should be resolved by #3865.
